[CI failure] org.elasticsearch.indices.stats.IndexStatsTests.testFieldDataStats

I observed this on the current query refactoring branch but it seems unrelated to the changes there.
I could also reproduce this on master.

http://build-us-00.elastic.co/job/es_feature_query_refactoring/2380

```
org.elasticsearch.indices.stats.IndexStatsTests.testFieldDataStats
Schlägt fehl seit 1 Build (Seit Fehlgeschlagen#2380 )
Dauer: 0.17 Sekunden.
Fehlermeldung

 Expected: <0L>      but: was <284L>

Stacktrace

java.lang.AssertionError: 
Expected: <0L>
     but: was <284L>
	at __randomizedtesting.SeedInfo.seed([B07CF80620C33000:A74387A13149B218]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.junit.Assert.assertThat(Assert.java:832)
	at org.elasticsearch.indices.stats.IndexStatsTests.testFieldDataStats(IndexStatsTests.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	[...] 
```




Tried this on master at 56c7cce4b and could also reproduce the same error with:

mvn clean test -Pdev -Dtests.seed=B07CF80620C33000 -Dtests.class=org.elasticsearch.indices.stats.IndexStatsTests -Dtests.method="testFieldDataStats" -Des.logger.level=DEBUG -Des.node.mode=local -Dtests.assertion.disabled=false -Dtests.security.manager=true -Dtests.heap.size=1024m -Dtests.locale=fr_FR -Dtests.timezone=Australia/Queensland

--- a/core/src/main/java/org/elasticsearch/indices/fielddata/cache/IndicesFieldDataCache.java
+++ b/core/src/main/java/org/elasticsearch/indices/fielddata/cache/IndicesFieldDataCache.java
@@ -35,7 +35,6 @@ import org.elasticsearch.index.fielddata.AtomicFieldData;
 import org.elasticsearch.index.fielddata.FieldDataType;
 import org.elasticsearch.index.fielddata.IndexFieldData;
 import org.elasticsearch.index.fielddata.IndexFieldDataCache;
-import org.elasticsearch.index.mapper.FieldMapper;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.shard.ShardId;
@@ -222,15 +221,22 @@ public class IndicesFieldDataCache extends AbstractComponent implements RemovalL
         @Override
         public void onClose(Object coreKey) {
             cache.invalidate(new Key(this, coreKey));
+            // don't call cache.cleanUp here as it would have bad performance implications
         }
 
         @Override
         public void onClose(IndexReader reader) {
             cache.invalidate(new Key(this, reader.getCoreCacheKey()));
+            // don't call cache.cleanUp here as it would have bad performance implications
         }
 
         @Override
         public void clear() {
+            for (Key key : cache.asMap().keySet()) {
+                if (key.indexCache.index.equals(index)) {
+                    cache.invalidate(key);
+                }
+            }
             // Note that cache invalidation in Guava does not immediately remove
             // values from the cache. In the case of a cache with a rare write or
             // read rate, it's possible for values to persist longer than desired.
@@ -238,11 +244,11 @@ public class IndicesFieldDataCache extends AbstractComponent implements RemovalL
             // Note this is intended by the Guava developers, see:
             // https://code.google.com/p/guava-libraries/wiki/CachesExplained#Eviction
             // (the "When Does Cleanup Happen" section)
-            for (Key key : cache.asMap().keySet()) {
-                if (key.indexCache.index.equals(index)) {
-                    cache.invalidate(key);
-                }
-            }
+
+            // We call it explicitly here since it should be a "rare" operation, and
+            // if a user runs it he probably wants to see memory returned as soon as
+            // possible
+            cache.cleanUp();
         }
 
         @Override
@@ -254,11 +260,16 @@ public class IndicesFieldDataCache extends AbstractComponent implements RemovalL
                     }
                 }
             }
+            // we call cleanUp() because this is a manual operation, should happen
+            // rarely and probably means the user wants to see memory returned as
+            // soon as possible
+            cache.cleanUp();
         }
 
         @Override
         public void clear(Object coreCacheKey) {
             cache.invalidate(new Key(this, coreCacheKey));
+            // don't call cache.cleanUp here as it would have bad performance implications
         }
     }
 

